contracts/m0/UIntMath.sol

```diff
@@ -0,0 +1,48 @@
+// SPDX-License-Identifier: GPL-3.0
+
+pragma solidity >=0.8.20 <0.9.0;
+
+/**
+ * @title  Library to perform safe math operations on uint types
+ * @author M0 Labs
+ */
+library UIntMath {
+    /* ============ Custom Errors ============ */
+
+    /// @notice Emitted when a passed value is greater than the maximum value of uint112.
+    error InvalidUInt112();
+
+    /// @notice Emitted when a passed value is greater than the maximum value of uint128.
+    error InvalidUInt128();
+
+    /* ============ Internal View/Pure Functions ============ */
+
+    /**
+     * @notice Casts a uint256 value to a uint112, ensuring that it is less than or equal to the maximum uint112 value.
+     * @param  n The value to cast.
+     * @return The value casted to uint112.
+     */
+    function safe112(uint256 n) internal pure returns (uint112) {
+        if (n > type(uint112).max) revert InvalidUInt112();
+        return uint112(n);
+    }
+
+    /**
+     * @notice Casts a uint256 value to a uint128, ensuring that it is less than or equal to the maximum uint128 value.
+     * @param  n The value to cast.
+     * @return The value casted to uint128.
+     */
+    function safe128(uint256 n) internal pure returns (uint128) {
+        if (n > type(uint128).max) revert InvalidUInt128();
+        return uint128(n);
+    }
+
+    /**
+     * @notice Limits a uint256 value to the maximum uint128 value.
+     * @param  n The value to bound.
+     * @return The value limited to within uint128 bounds.
+     */
+    function bound128(uint256 n) internal pure returns (uint128) {
+        return uint128(min256(n, uint256(type(uint128).max)));
```

Comment:
Looks like we're missing the definition of [`min256`](https://github.com/m0-foundation/common/blob/66657d984bb77ba83863dd4607704594a6c610bc/src/libs/UIntMath.sol#L221-L229) 👀